### DISCUSSION TITLE
Feature: Automatically deploy branches that match pattern

### DISCUSSION THREAD

**Discussion by nstogner** [fluxcd/flux2#3084] [Proposals]
The use-case I would like to solve with FluxCD is:

For each branch that is created in git that matches some pattern (`feature/*`), I would like to be able to template out a `kind: Kustomization` (possibly also a `kind: HelmRelease`) that will deploy that branch. When the branch gets deleted I would like the Kustomization/HelmRelease to be deleted as well.

**Comment by stefanprodan** [comment]
You could do this with GitHub Actions or any other CI system that knows how to react to branch events. On branch creation, the CI job will generate the GitRepository+Kustomization YAMLs, push those to the bootstrap repo under `clusters/preview/<branch-name>` and Flux on the preview cluster will pick it up. On branch delete, the CI job will delete the `clusters/preview/<branch-name>` directory from the bootstrap repo.

**Reply by nstogner** [reply]
Agreed, I think that approach makes the most sense with current functionality. However, it seems like the requested functionality would fit nicely within the flux controllers which cover the use-case of watching git and interacting with the k8s api (`git --> k8s api`). The approach of `git --actions--> git --flux--> k8s api` is less desirable as we prefer to minimize automated commits because they tend to cause complexity when mixed with human authored commits.

**Reply by nstogner** [reply]
I would be interested in your take on whether this functionality belongs in Flux.

**Reply by stefanprodan** [reply]
A preview controller is something that we're considering, see https://github.com/fluxcd/flux2/discussions/831#discussioncomment-323218

Adding more controllers and APIs to Flux is not something that I would consider for this year. Our focus is to harden the current APIs and release Flux v2 GA https://fluxcd.io/roadmap/#the-road-to-flux-v2-ga
